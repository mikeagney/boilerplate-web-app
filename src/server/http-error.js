class HttpError extends Error {
  constructor(message, { status = 500, headers = {} } = {}) {
    super(message);
    this.status = status;
    this.headers = headers;
  }
}

export default HttpError;
