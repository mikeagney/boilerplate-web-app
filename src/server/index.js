import '@babel/polyfill';
import App from './app';

const app = new App();
app.initialize().listen();
