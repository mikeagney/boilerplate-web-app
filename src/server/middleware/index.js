import { logger, errorLogger } from './logger-middlewares';

/**
 * Gets the Express middlewares that should be applied before any
 * application-specific handlers.
 * @returns {import('express').Handler[]}
 */
export function preAppHandlers() {
  return [
    logger(),
  ];
}

/**
 * Gets the Express middlewares that should be applied after any
 * application-specific handlers.
 * @returns {import('express').Handler[]}
 */
export function postAppHandlers() {
  return [
    errorLogger(),
  ];
}
