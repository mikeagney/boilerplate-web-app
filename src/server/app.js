import express from 'express';
import retry from 'async-retry';
import config from '../common/config';
import { server as serverLogger } from '../common/logger';
import DbMigrate from './client/db-migrate';
import { preAppHandlers, postAppHandlers } from './middleware';
import RootRouter from './paths/root-router';

class App {
  initialize() {
    this.app = express();
    this.app.use(...preAppHandlers());
    this.app.use(new RootRouter().initialize().router);
    this.app.use(...postAppHandlers());
    this.config = config();
    this.logger = serverLogger();
    this.dbMigrate = new DbMigrate();
    return this;
  }

  async listen() {
    const { name, server } = this.config;
    const { port } = server;

    await retry(() => this.dbMigrate.up());

    this.app.listen(port, () =>
      this.logger.info(
        `Example app environment "${name}" listening on port ${port} from dir ${__dirname}!`,
      ));
  }
}

export default App;
