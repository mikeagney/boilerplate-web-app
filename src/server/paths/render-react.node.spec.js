import { readFile } from 'fs';
import path from 'path';
import React from 'react';
import { ChunkExtractor } from '@loadable/server';
import { renderToString } from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Routes from '../../client/pages/routes';
import createStore from '../../client/store';
import RenderReact from './render-react';

jest
  .mock('fs')
  .mock('path')
  .mock('@loadable/server')
  .mock('react-dom/server')
  .mock('react-router-dom')
  .mock('../../client/pages/routes')
  .mock('../../client/store');

describe('RenderReact route', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('getTemplate', () => {
    it('will return the template from the expected path', () => {
      // Arrange
      const indexPath = 'root/dist/client/templates/index.html';
      path.resolve.mockReturnValue(indexPath);

      // Act
      RenderReact.getTemplate();

      // Assert
      expect(path.resolve).toHaveBeenCalledWith(
        expect.any(String),
        '../../../dist/client/templates/index.html',
      );
      expect(readFile).toHaveBeenCalledWith(indexPath, { encoding: 'utf8' }, expect.any(Function));
    });

    it('will resolve with the expected content when file read succeeds', async () => {
      // Arrange
      const templateResult = 'Template Result';
      const templatePromise = RenderReact.getTemplate();
      const readFileCallback = readFile.mock.calls[0][2];

      // Act
      readFileCallback(null, templateResult);

      // Assert
      expect(templatePromise).resolves.toEqual(templateResult);
    });

    it('will reject with the expected error when file read fails', async () => {
      // Arrange
      const templateError = new Error('File read failed');
      const templatePromise = RenderReact.getTemplate();
      const readFileCallback = readFile.mock.calls[0][2];

      // Act
      readFileCallback(templateError);

      // Assert
      expect(templatePromise).rejects.toBe(templateError);
    });
  });

  describe('renderApp', () => {
    // eslint-disable-next-line react/prefer-stateless-function
    class MockStaticRouter extends React.Component {
      render() {
        const { children, ...otherProps } = this.props;
        return (
          <div {...otherProps}>
            Mock static router
            <>{children}</>
          </div>
        );
      }
    }
    const mockChunkExtractor = {
      collectChunks: jest.fn(),
      getScriptTags: jest.fn(),
    };

    beforeEach(() => {
      StaticRouter.mockImplementation(props => new MockStaticRouter(props));
      jest.spyOn(RenderReact, 'createStore');
    });

    afterEach(() => {
      RenderReact.createStore.mockRestore();
    });

    it('will render the expected app', () => {
      // Arrange
      const statsFile = 'root/dist/client/loadable-stats.json';
      const initialState = { mockState: true };
      const store = { getState: jest.fn(() => initialState) };
      RenderReact.createStore.mockReturnValue(store);
      path.resolve.mockReturnValue(statsFile);
      renderToString.mockReturnValue('Mock app');
      ChunkExtractor.mockImplementation(() => mockChunkExtractor);
      mockChunkExtractor.collectChunks.mockReturnValue(<div>Mock rendered app</div>);
      mockChunkExtractor.getScriptTags.mockReturnValue('script tags');

      // Act
      const result = RenderReact.renderApp({ url: '/' });

      // Assert
      expect(result).toEqual({
        context: {},
        content: 'Mock app',
        scriptTags: 'script tags',
        initialState: JSON.stringify(initialState),
      });
      expect(ChunkExtractor).toHaveBeenCalledWith({ statsFile });
      expect(renderToString).toHaveBeenCalledWith(expect.anything());
      const renderedApp = mockChunkExtractor.collectChunks.mock.calls[0][0];
      expect(toJson(shallow(renderedApp))).toMatchSnapshot();
    });
  });

  describe('createStore', () => {
    it('will call client createStore', () => {
      // Arrange
      const store = {};
      createStore.mockReturnValue(store);

      // Act
      const result = RenderReact.createStore();

      // Assert
      expect(result).toBe(store);
    });
  });

  describe('route', () => {
    const getTemplate = jest.fn();
    const renderApp = jest.fn();

    beforeEach(() => {
      jest.spyOn(RenderReact, 'getTemplate').mockImplementation(getTemplate);
      jest.spyOn(RenderReact, 'renderApp').mockImplementation(renderApp);
    });

    afterEach(() => {
      RenderReact.getTemplate.mockRestore();
      RenderReact.renderApp.mockRestore();
    });

    it('will not render for unsupported paths', async () => {
      // Arrange
      const req = {};
      const res = { send: jest.fn() };
      const next = jest.fn();
      Routes.getMatchingRoute.mockReturnValue(null);

      // Act
      await RenderReact.route(req, res, next);

      // Assert
      expect(next).toHaveBeenCalledTimes(1);
      expect(res.send).not.toHaveBeenCalled();
    });

    it('will redirect if the component renders as a redirect', async () => {
      // Arrange
      getTemplate.mockResolvedValue('Html template with content "{content}{scriptTags}" here');
      renderApp.mockReturnValue({
        context: { url: '/redirect-url' },
        content: 'Mock content',
        scriptTags: ' and mock script tags',
      });

      const req = { path: '/' };
      const res = { send: jest.fn(), redirect: jest.fn() };
      const next = jest.fn();
      Routes.getMatchingRoute.mockReturnValue({});

      // Act
      await RenderReact.route(req, res, next);

      // Assert
      expect(next).not.toHaveBeenCalled();
      expect(res.send).not.toHaveBeenCalled();
      expect(res.redirect).toHaveBeenCalledWith('/redirect-url');
    });

    it('will load the template, render the app, and return the expected result', async () => {
      // Arrange
      getTemplate.mockResolvedValue('Html template with content "{content}{scriptTags}" here');
      renderApp.mockReturnValue({
        context: {},
        content: 'Mock content',
        scriptTags: ' and mock script tags',
      });

      const req = { path: '/' };
      const res = { send: jest.fn() };
      const next = jest.fn();
      Routes.getMatchingRoute.mockReturnValue({});

      // Act
      await RenderReact.route(req, res, next);

      // Assert
      expect(next).not.toHaveBeenCalled();
      expect(res.send).toHaveBeenCalledWith(
        'Html template with content "Mock content and mock script tags" here',
      );
    });
  });
});
