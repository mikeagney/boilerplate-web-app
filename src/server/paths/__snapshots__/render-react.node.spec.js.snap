// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderReact route renderApp will render the expected app 1`] = `
<div
  context={
    Object {
      "mockContext": "mock",
    }
  }
  location="/"
>
  Mock static router
  <App />
</div>
`;
